### ISSUE TITLE
Resource Request: Ohai Resource (system information)

### ISSUE THREAD

**Issue by clintoncwolfe** [inspec/inspec#3886]
### 🎛 Description

It is common to want to have available information about the target machine, such as networking information, detailed OS version, and other similar information.

The Chef project [`ohai`](https://github.com/chef/ohai) provides exactly that, in a modular fashion. Create an InSpec resource that can be used as an informational resource to query aspects of the target environment.

Consider loading properties lazily, to take advantage of ohai's modularity.

### 🌍 InSpec and Platform Version

3.7.x

### 💁 Possible Solutions

```

the_fqdn = ohai.networking.fqdn

```

**Comment by aaronlippold** [comment]
I would be happy to help with this. Would this inherit from the JSON resource base?

**Comment by burtlo** [comment]
Here is some starting code for that solution: https://github.com/burtlo/inspec-ohai_resource/blob/master/ohai/libraries/ohai.rb

* It can't rely on loading the ohai library inside of Ruby without some serious error handling as ohai and inspec load different versions of mixlib libraries that are direct conflict with one another.

* It isn't platform independent. It uses `inspec.command('which ohai')` to find the ohai command to execute. I'm sure that could easily be modified or that value could be hard-coded to assume someone has Chef installed via the ChefDK. Perhaps changed instead to check for presence of file.

* It doesn't load things in a modular way but it does cache the content once its loaded.

* It also currently doesn't disable some common extensions that are taken care of by other resources within inspec (e.g. etc, os) that would decrease the ohai execution time.

* It does provide the dot notation interface throughout the entire result-set by creating methods for top-level keys and then relying on `Hashie::Mash`.

* It overrides `Hashie::Mash` to disable warnings because there are some ohai result properties that collide with `Hashie::Mash` default methods. I think this is fine as this is really read-only.

**Comment by aaronlippold** [comment]
We are very interested in moving forward with this. After some conversations out of band, I think we need to decide if we want to push this into train. The ideal situation would be to not require ohai on the target if possible - which I think means we need to do this via train. What are folks thoughts? If we can define a 'first cut' I can put some resources on it. @burtlo made a good point that perhaps we don't need all of Ohai and if the first cut was just the core subset of the data we want to hit first.

**Comment by burtlo** [comment]
I think an ohai resource would be a nice thing to have and the example implementation I provided above is a decent starting point. What you are after sounds like something entirely different to me. I'll summarize things as I understand it:

@aaronlippold you mentioned to me that ideally you would be able to initiate a scan and gain additional information about the system, akin to what you would capture with Ohai. You specifically mentioned that you would prefer that the ohai script/binary is not present on the system. I shared with you that I believe Ohai is implemented in a way that it assumes it is present on the system (e.g. `File.read`, `shell_out`). I proposed a few alternatives:

* Write a ruby script that uses train to invoke commands on the remote instance to capture the information you want from the target. Perform a scan against the target. Merge the results. (minor)

* Re-write the Ohai gem so that it used a layer of indirection (e.g. train) when it accessed the system so that a remote target could be provided. (epic)

@aaronlippold suggested pass-through attributes that could be assigned to scans. You also talked about `inspec detect` expanded to capturing more data.
